## Problem Statement

The IPL is a fast-paced T20 league, and understanding match trends can be complex. This project analyzes **ball-by-ball data (2008-2023)** to explore **batting, bowling, match results, and venue impact** using **Exploratory Data Analysis (EDA)** to uncover key insights.

## **Dataset Info:**

**Match ID** : Unique identifier for each match.

**Date**: Date on which the match was played.

**Venue** : Name of the stadium where the match was played.

**Bat First** : Name of the team that batted first.

**Bat Second** : Name of the team that batted second.

**Innings** : Indicates the inning number (1 or 2) of the match.

**Over** : Specifies the over number within the inning.

**Ball** : Denotes the ball number within the over.

**Batter** : Name of the batsman facing the delivery.

**Non Striker** : Name of the non-striking batsman.

**Bowler** : Name of the bowler delivering the ball.

**Batter Runs** : Runs scored by the batsman on that specific delivery.

**Extra Runs** : Additional runs awarded due to extras like wides or no-balls.

**Runs From Ball** : Total runs scored from that delivery (batsman + extras).

**Ball Rebowled** : Indicator if the ball was rebowled (e.g., after a no-ball).

**Extra Type** : Type of extra run conceded (e.g., wide, no-ball, leg bye, bye).

**Wicket** : Indicates if a wicket fell on that delivery (1 for yes, 0 for no).

**Method** : Type of dismissal (e.g., bowled, caught, run out, etc.).

**Player Out** : Name of the dismissed player, if applicable.

**Innings Runs** : Total runs scored in the inning up to that ball.

**Innings Wickets** : Total number of wickets lost in the inning up to that ball.

**Target Score** : The target score set by the batting team (useful in chases).

**Runs to Get** : Runs required to win the match (only in chases).

**Balls Remaining** : Number of balls left in the innings.

**Winner** : Name of the team that won the match.

**Chased Successfully** : Indicator if the target was successfully chased (1 for yes, 0 for no).

**Total Batter Runs** : The total runs accumulated by the batter up until this delivery in the innings

**Total Non Striker Runs** : The total runs accumulated by the non-striker up until this delivery in the innings

**Batter Balls Faced** : The number of valid deliveries faced by the batter up until this delivery in the innings.

**Non Striker Balls Faced** : The number of valid deliveries faced by the non-striker up until this delivery in the innings.

**Player Out Runs** : The total runs scored by the dismissed player before getting out

**Player Out Balls Faced** : The number of valid deliveries faced by the dismissed player before getting out

**Bowler Runs Conceded** : Total runs conceded by the bowler in that ball.

**Valid Ball** : Indicates if the delivery was a valid ball (not wide/no-ball).

## **Import Libraries**

In [9]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

## **Load Data**

In [10]:
df = pd.read_csv('/content/ball_by_ball_ipl.csv', index_col = 0)

### **Dataset Overview & Structure**

In [11]:
df.head(10)

Match ID        Date         Venue            Bat First  \
0   1359507  2023-04-23  Eden Gardens  Chennai Super Kings   
1   1359507  2023-04-23  Eden Gardens  Chennai Super Kings   
2   1359507  2023-04-23  Eden Gardens  Chennai Super Kings   
3   1359507  2023-04-23  Eden Gardens  Chennai Super Kings   
4   1359507  2023-04-23  Eden Gardens  Chennai Super Kings   
5   1359507  2023-04-23  Eden Gardens  Chennai Super Kings   
6   1359507  2023-04-23  Eden Gardens  Chennai Super Kings   
7   1359507  2023-04-23  Eden Gardens  Chennai Super Kings   
8   1359507  2023-04-23  Eden Gardens  Chennai Super Kings   
9   1359507  2023-04-23  Eden Gardens  Chennai Super Kings   

              Bat Second  Innings  Over  Ball      Batter Non Striker  ...  \
0  Kolkata Knight Riders        1     1     1  RD Gaikwad   DP Conway  ...   
1  Kolkata Knight Riders        1     1     2  RD Gaikwad   DP Conway  ...   
2  Kolkata Knight Riders        1     1     3  RD Gaikwad   DP Conway  ...   
3  Kolkata Knight Riders        1     1     4  RD Gaikwad   DP Conway  ...   
4  Kolkata Knight Riders        1     1     5  RD Gaikwad   DP Conway  ...   
5  Kolkata Knight Riders        1     1     6   DP Conway  RD Gaikwad  ...   
6  Kolkata Knight Riders        1     2     1  RD Gaikwad   DP Conway  ...   
7  Kolkata Knight Riders        1     2     2   DP Conway  RD Gaikwad  ...   
8  Kolkata Knight Riders        1     2     3   DP Conway  RD Gaikwad  ...   
9  Kolkata Knight Riders        1     2     4  RD Gaikwad   DP Conway  ...   

                Winner  Chased Successfully  Total Batter Runs  \
0  Chennai Super Kings                    0                  4   
1  Chennai Super Kings                    0                  4   
2  Chennai Super Kings                    0                  4   
3  Chennai Super Kings                    0                  4   
4  Chennai Super Kings                    0                  5   
5  Chennai Super Kings                    0                  0   
6  Chennai Super Kings                    0                  6   
7  Chennai Super Kings                    0                  0   
8  Chennai Super Kings                    0                  1   
9  Chennai Super Kings                    0                  7   

   Total Non Striker Runs  Batter Balls Faced Non Striker Balls Faced  \
0                       0                   1                       0   
1                       0                   2                       0   
2                       0                   3                       0   
3                       0                   4                       0   
4                       0                   5                       0   
5                       5                   1                       5   
6                       0                   6                       1   
7                       6                   2                       6   
8                       6                   3                       6   
9                       1                   7                       3   

   Player Out Runs Player Out Balls Faced Bowler Runs Conceded  Valid Ball  
0              NaN                    NaN                    4           1  
1              NaN                    NaN                    0           1  
2              NaN                    NaN                    0           1  
3              NaN                    NaN                    0           1  
4              NaN                    NaN                    1           1  
5              NaN                    NaN                    0           1  
6              NaN                    NaN                    1           1  
7              NaN                    NaN                    0           1  
8              NaN                    NaN                    1           1  
9              NaN                    NaN                    1           1  

[10 rows x 34 columns]

In [12]:
df.tail()

Match ID        Date                               Venue  \
239688   1136596  2018-05-05  Rajiv Gandhi International Stadium   
239689   1136596  2018-05-05  Rajiv Gandhi International Stadium   
239690   1136596  2018-05-05  Rajiv Gandhi International Stadium   
239691   1136596  2018-05-05  Rajiv Gandhi International Stadium   
239692   1136596  2018-05-05  Rajiv Gandhi International Stadium   

               Bat First           Bat Second  Innings  Over  Ball  \
239688  Delhi Daredevils  Sunrisers Hyderabad        2    20     1   
239689  Delhi Daredevils  Sunrisers Hyderabad        2    20     2   
239690  Delhi Daredevils  Sunrisers Hyderabad        2    20     3   
239691  Delhi Daredevils  Sunrisers Hyderabad        2    20     4   
239692  Delhi Daredevils  Sunrisers Hyderabad        2    20     5   

               Batter    Non Striker  ...               Winner  \
239688      YK Pathan  KS Williamson  ...  Sunrisers Hyderabad   
239689      YK Pathan  KS Williamson  ...  Sunrisers Hyderabad   
239690      YK Pathan  KS Williamson  ...  Sunrisers Hyderabad   
239691      YK Pathan  KS Williamson  ...  Sunrisers Hyderabad   
239692  KS Williamson      YK Pathan  ...  Sunrisers Hyderabad   

        Chased Successfully  Total Batter Runs  Total Non Striker Runs  \
239688                    1                 16                      31   
239689                    1                 22                      31   
239690                    1                 26                      31   
239691                    1                 27                      31   
239692                    1                 32                      27   

        Batter Balls Faced Non Striker Balls Faced  Player Out Runs  \
239688                   9                      29              NaN   
239689                  10                      29              NaN   
239690                  11                      29              NaN   
239691                  12                      29              NaN   
239692                  30                      12              NaN   

       Player Out Balls Faced Bowler Runs Conceded  Valid Ball  
239688                    NaN                    2           1  
239689                    NaN                    6           1  
239690                    NaN                    4           1  
239691                    NaN                    1           1  
239692                    NaN                    1           1  

[5 rows x 34 columns]

In [13]:
df.shape

(239693, 34)

#### **Identifying key columns and understanding data types**

In [14]:
df.columns

Index(['Match ID', 'Date', 'Venue', 'Bat First', 'Bat Second', 'Innings',
       'Over', 'Ball', 'Batter', 'Non Striker', 'Bowler', 'Batter Runs',
       'Extra Runs', 'Runs From Ball', 'Ball Rebowled', 'Extra Type', 'Wicket',
       'Method', 'Player Out', 'Innings Runs', 'Innings Wickets',
       'Target Score', 'Runs to Get', 'Balls Remaining', 'Winner',
       'Chased Successfully', 'Total Batter Runs', 'Total Non Striker Runs',
       'Batter Balls Faced', 'Non Striker Balls Faced', 'Player Out Runs',
       'Player Out Balls Faced', 'Bowler Runs Conceded', 'Valid Ball'],
      dtype='object')

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 239693 entries, 0 to 239692
Data columns (total 34 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Match ID                 239693 non-null  int64  
 1   Date                     239693 non-null  object 
 2   Venue                    239693 non-null  object 
 3   Bat First                239693 non-null  object 
 4   Bat Second               239693 non-null  object 
 5   Innings                  239693 non-null  int64  
 6   Over                     239693 non-null  int64  
 7   Ball                     239693 non-null  int64  
 8   Batter                   239693 non-null  object 
 9   Non Striker              239693 non-null  object 
 10  Bowler                   239693 non-null  object 
 11  Batter Runs              239693 non-null  int64  
 12  Extra Runs               239693 non-null  int64  
 13  Runs From Ball           239693 non-null  int64  
 14  Ball Rebo

In [16]:
df.dtypes

Match ID                     int64
Date                        object
Venue                       object
Bat First                   object
Bat Second                  object
Innings                      int64
Over                         int64
Ball                         int64
Batter                      object
Non Striker                 object
Bowler                      object
Batter Runs                  int64
Extra Runs                   int64
Runs From Ball               int64
Ball Rebowled                int64
Extra Type                  object
Wicket                       int64
Method                      object
Player Out                  object
Innings Runs                 int64
Innings Wickets              int64
Target Score                 int64
Runs to Get                float64
Balls Remaining              int64
Winner                      object
Chased Successfully          int64
Total Batter Runs            int64
Total Non Striker Runs       int64
Batter Balls Faced           int64
Non Striker Balls Faced      int64
Player Out Runs            float64
Player Out Balls Faced     float64
Bowler Runs Conceded         int64
Valid Ball                   int64
dtype: object

In [17]:
df.size #it will gives non nan or null element

8149562

**DATA CLEANING**

**Standardizing Categorical values**

Before moving to analysis, ensure data consistency in categorical values:
Checking key categorical columns prevents errors in grouping, filtering, and aggregations


In [18]:
categorical_columns = df.select_dtypes(include= ['object','category']).columns.tolist()

In [19]:
df['Venue'].unique()

array(['Eden Gardens', 'De Beers Diamond Oval', 'Sharjah Cricket Stadium',
       'Wankhede Stadium', 'Feroz Shah Kotla', 'Holkar Cricket Stadium',
       'Punjab Cricket Association IS Bindra Stadium',
       'Maharashtra Cricket Association Stadium',
       'MA Chidambaram Stadium', 'Dubai International Cricket Stadium',
       'Kingsmead', 'Dr DY Patil Sports Academy', 'SuperSport Park',
       'Sawai Mansingh Stadium',
       'Shaheed Veer Narayan Singh International Stadium',
       'M Chinnaswamy Stadium', 'Rajiv Gandhi International Stadium',
       'Brabourne Stadium', 'Sheikh Zayed Stadium',
       'M.Chinnaswamy Stadium', 'Punjab Cricket Association Stadium',
       'Narendra Modi Stadium', 'Arun Jaitley Stadium',
       'Barabati Stadium', 'Saurashtra Cricket Association Stadium',
       'Sardar Patel Stadium',
       'Himachal Pradesh Cricket Association Stadium', 'Nehru Stadium',
       'Green Park', 'Subrata Roy Sahara Stadium',
       'JSCA International Stadium Complex'

In [20]:
venue_map = {
    "M.Chinnaswamy Stadium": "M Chinnaswamy Stadium",
    "Sardar Patel Stadium": "Narendra Modi Stadium",
    "Punjab Cricket Association IS Bindra Stadium": "Punjab Cricket Association Stadium",
    "Feroz Shah Kotla": "Arun Jaitley Stadium",
    "Zayed Cricket Stadium": "Sheikh Zayed Stadium",
    "De Beers Diamond Oval": "Kimberley Oval",
    "Rajiv Gandhi International Stadium": "Uppal Stadium",
    "Dr Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium": "ACA-VDCA Stadium",
    "Shaheed Veer Narayan Singh International Stadium": "Raipur International Cricket Stadium",
    "JSCA International Stadium Complex": "Ranchi Stadium",
    "Barsapara Cricket Stadium": "Assam Cricket Association Stadium"
}

df['venue'] = df['Venue'].replace(venue_map)

In [21]:
df['Bat First'].unique()

array(['Chennai Super Kings', 'Rajasthan Royals', 'Mumbai Indians',
       'Delhi Daredevils', 'Royal Challengers Bangalore',
       'Delhi Capitals', 'Sunrisers Hyderabad', 'Rising Pune Supergiants',
       'Punjab Kings', 'Kings XI Punjab', 'Kochi Tuskers Kerala',
       'Pune Warriors', 'Deccan Chargers', 'Kolkata Knight Riders',
       'Gujarat Titans', 'Rising Pune Supergiant', 'Gujarat Lions',
       'Lucknow Super Giants'], dtype=object)

In [22]:
df['Bat Second'].unique()

array(['Kolkata Knight Riders', 'Chennai Super Kings', 'Delhi Capitals',
       'Kings XI Punjab', 'Kochi Tuskers Kerala', 'Delhi Daredevils',
       'Mumbai Indians', 'Gujarat Titans', 'Rajasthan Royals',
       'Royal Challengers Bangalore', 'Deccan Chargers',
       'Sunrisers Hyderabad', 'Lucknow Super Giants', 'Punjab Kings',
       'Pune Warriors', 'Gujarat Lions', 'Rising Pune Supergiants',
       'Rising Pune Supergiant'], dtype=object)

In [23]:
df['Winner'].unique()

array(['Chennai Super Kings', 'Delhi Capitals', 'Mumbai Indians',
       'Kings XI Punjab', 'Kochi Tuskers Kerala',
       'Royal Challengers Bangalore', 'Gujarat Titans', 'Punjab Kings',
       'Kolkata Knight Riders', 'Delhi Daredevils', 'Pune Warriors',
       'Rajasthan Royals', 'Sunrisers Hyderabad', 'Lucknow Super Giants',
       'Deccan Chargers', 'Gujarat Lions', 'Rising Pune Supergiants',
       'Rising Pune Supergiant'], dtype=object)

In [24]:
team_name_map = {
    "Delhi Daredevils": "Delhi Capitals",
    "Kings XI Punjab": "Punjab Kings"
}
df['Bat First'] = df['Bat First'].replace(team_name_map)
df['Bat Second'] = df['Bat Second'].replace(team_name_map)
df['Winner'] = df['Winner'].replace(team_name_map)

**CLEANING PLAYER NAMES**

In [25]:
df['Batter'].unique()

array(['RD Gaikwad', 'DP Conway', 'AM Rahane', 'S Dube', 'RA Jadeja',
       'MS Dhoni', 'N Jagadeesan', 'SP Narine', 'VR Iyer', 'N Rana',
       'JJ Roy', 'RK Singh', 'AD Russell', 'D Wiese', 'UT Yadav',
       'CV Varun', 'GC Smith', 'NV Ojha', 'SA Asnodkar', 'YK Pathan',
       'LA Carseldine', 'AS Raut', 'SK Warne', 'SM Harwood', 'M Vijay',
       'ML Hayden', 'SK Raina', 'S Badrinath', 'JDP Oram', 'RG Sharma',
       'Q de Kock', 'SA Yadav', 'SS Tiwary', 'KA Pollard', 'HH Pandya',
       'KH Pandya', 'NM Coulter-Nile', 'J Yadav', 'JJ Bumrah', 'PP Shaw',
       'S Dhawan', 'SPD Smith', 'RR Pant', 'SS Iyer', 'AR Patel',
       'SO Hetmyer', 'R Ashwin', 'LMP Simmons', 'PA Patel', 'AT Rayudu',
       'Harbhajan Singh', 'J Suchith', 'DR Smith', 'F du Plessis',
       'MEK Hussey', 'DJ Bravo', 'P Negi', 'MM Sharma', 'A Nehra',
       'G Gambhir', 'V Sehwag', 'TM Dilshan', 'MF Maharoof',
       'Shoaib Malik', 'SE Marsh', 'JR Hopes', 'Yuvraj Singh',
       'LA Pomersbach', 'DPMD Jayaward

In [26]:
df['Batter'] = df['Batter'].str.strip().str.title() #str.strip removes the spaces between starting and ending of a string
#title makes the first letter as capital in each word
df['Non Striker'] = df['Non Striker'].str.strip().str.title()
df['Bowler'] = df['Bowler'].str.strip().str.title()
df['Player Out'] = df['Player Out'].str.strip().str.title()

**HANDLING MISSING & DUPLICATED DATA**

In [27]:
df.info() #check the method and some other doesnt have values

<class 'pandas.core.frame.DataFrame'>
Index: 239693 entries, 0 to 239692
Data columns (total 35 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Match ID                 239693 non-null  int64  
 1   Date                     239693 non-null  object 
 2   Venue                    239693 non-null  object 
 3   Bat First                239693 non-null  object 
 4   Bat Second               239693 non-null  object 
 5   Innings                  239693 non-null  int64  
 6   Over                     239693 non-null  int64  
 7   Ball                     239693 non-null  int64  
 8   Batter                   239693 non-null  object 
 9   Non Striker              239693 non-null  object 
 10  Bowler                   239693 non-null  object 
 11  Batter Runs              239693 non-null  int64  
 12  Extra Runs               239693 non-null  int64  
 13  Runs From Ball           239693 non-null  int64  
 14  Ball Rebo

In [28]:
df.isna().sum() #sum will see the True where is present if true means it will sum it and it will represent the sum of missing/empty/undefined value in the dataset

Match ID                        0
Date                            0
Venue                           0
Bat First                       0
Bat Second                      0
Innings                         0
Over                            0
Ball                            0
Batter                          0
Non Striker                     0
Bowler                          0
Batter Runs                     0
Extra Runs                      0
Runs From Ball                  0
Ball Rebowled                   0
Extra Type                      0
Wicket                          0
Method                     227866
Player Out                 227866
Innings Runs                    0
Innings Wickets                 0
Target Score                    0
Runs to Get                123927
Balls Remaining                 0
Winner                          0
Chased Successfully             0
Total Batter Runs               0
Total Non Striker Runs          0
Batter Balls Faced              0
Non Striker Balls Faced         0
Player Out Runs            227866
Player Out Balls Faced     227866
Bowler Runs Conceded            0
Valid Ball                      0
venue                           0
dtype: int64

**OBSERVATIONS**

* **Method** has **227,866 missing values** because it is only applicable when a wicket falls. It is **NaN for deliveries where no wicket fell**.  
* **Player Out** has **227,866 missing values** as it is recorded only when a wicket falls, making it **NaN for non-wicket deliveries**.  
* **Player Out Runs** has **227,866 missing values** because it is missing for deliveries where no player was out.  
* **Player Out Balls Faced** has **227,866 missing values** for the same reason—it is recorded only when a player gets out.  
* **Runs to Get** has **123,927 missing values** as it is **only relevant for the second innings**, making it **NaN for the first innings**.

In [29]:
df.dropna()

Match ID        Date                               Venue  \
126      1359507  2023-04-23                        Eden Gardens   
130      1359507  2023-04-23                        Eden Gardens   
167      1359507  2023-04-23                        Eden Gardens   
174      1359507  2023-04-23                        Eden Gardens   
213      1359507  2023-04-23                        Eden Gardens   
...          ...         ...                                 ...   
239434    598012  2013-04-13              MA Chidambaram Stadium   
239438    598012  2013-04-13              MA Chidambaram Stadium   
239624   1136596  2018-05-05  Rajiv Gandhi International Stadium   
239637   1136596  2018-05-05  Rajiv Gandhi International Stadium   
239675   1136596  2018-05-05  Rajiv Gandhi International Stadium   

                          Bat First             Bat Second  Innings  Over  \
126             Chennai Super Kings  Kolkata Knight Riders        2     1   
130             Chennai Super Kings  Kolkata Knight Riders        2     2   
167             Chennai Super Kings  Kolkata Knight Riders        2     8   
174             Chennai Super Kings  Kolkata Knight Riders        2     9   
213             Chennai Super Kings  Kolkata Knight Riders        2    15   
...                             ...                    ...      ...   ...   
239434  Royal Challengers Bangalore    Chennai Super Kings        2    19   
239438  Royal Challengers Bangalore    Chennai Super Kings        2    19   
239624               Delhi Capitals    Sunrisers Hyderabad        2     9   
239637               Delhi Capitals    Sunrisers Hyderabad        2    11   
239675               Delhi Capitals    Sunrisers Hyderabad        2    18   

        Ball        Batter    Non Striker  ... Chased Successfully  \
126        4     Sp Narine   N Jagadeesan  ...                   0   
130        2  N Jagadeesan        Vr Iyer  ...                   0   
167        1       Vr Iyer         N Rana  ...                   0   
174        2        N Rana         Jj Roy  ...                   0   
213        3        Jj Roy       Rk Singh  ...                   0   
...      ...           ...            ...  ...                 ...   
239434     1      Ms Dhoni      Ra Jadeja  ...                   1   
239438     4      Dj Bravo      Ra Jadeja  ...                   1   
239624     6      Ad Hales       S Dhawan  ...                   1   
239637     6      S Dhawan  Ks Williamson  ...                   1   
239675     1     Mk Pandey  Ks Williamson  ...                   1   

        Total Batter Runs  Total Non Striker Runs  Batter Balls Faced  \
126                     0                       1                   0   
130                     0                       0                   0   
167                     0                      22                   0   
174                     0                      19                   0   
213                     0                      21                   0   
...                   ...                     ...                 ...   
239434                  0                      26                   0   
239438                  0                      26                   0   
239624                  0                      28                   0   
239637                  0                       4                   0   
239675                  0                      28                   0   

        Non Striker Balls Faced Player Out Runs  Player Out Balls Faced  \
126                           1             0.0                     3.0   
130                           2             1.0                     3.0   
167                          18            20.0                    20.0   
174                           4            27.0                    21.0   
213                          16            61.0                    27.0   
...                         ...             ...                     ...   
239434                 

In [30]:
df['Method'].fillna('Not Out', inplace = True) #here inplace = True will modify the orginal column/series
df['Player Out'].fillna('None', inplace = True) #if the player is not out the value will be empty instaed of it we will replace with None
df['Player Out Runs'].fillna(0, inplace = True) #if that particular ball is the player is not out we will put 0
df['Player Out Balls Faced'].fillna(0, inplace = True)


**CHECK FOR DUPLICATES**

In [31]:
df.duplicated().sum()#check the each row if the row is seen before in the dataset return true else false

np.int64(0)

**DATA TYPE CORRECTIONS**

**Match ID** should be converted from **int64** to **str** because it represents a categorical value, not a numerical one.

**Date** should be converted from **object** to **datetime64** since it is needed for time-series analysis.

**Ball Reboweled** should be converted from **int64** to **bool** by mapping 1 → True and 0 → False.

**Runs to Get** should be converted from **float64** to **int64** because runs are always whole numbers.

**Player Out Runs** should be converted from **object** to **int64** because it should be a numeric column.

**Player Out Balls** Faced should be converted from **float64** to **int64** because it should be an integer.

In [32]:
df['Match ID'] = df['Match ID'].astype(str)
df['Ball Rebowled'] = df['Ball Rebowled'].astype(bool)
df['Runs to Get'].fillna(0, inplace=True)
df['Runs to Get'] = df['Runs to Get'].astype(int)
df['Player Out Runs'].fillna(0, inplace=True)
df['Player Out Runs'] = df['Player Out Runs'].astype(int)
df['Player Out Balls Faced'].fillna(0, inplace=True)
df['Player Out Balls Faced'] = df['Player Out Balls Faced'].astype(int)
df['Date'] = pd.to_datetime(df['Date'])

**FEATURE ENGINEERING**

In [36]:
df['Year'] = pd.to_datetime(df['Date']).dt.year

In [41]:
df['is_six'] = (df['Batter Runs'] == 6 ).astype(int) #the true or false will be converted to 0 or 1 by using this astype(int)

In [43]:
df['Method'].unique()

array(['Not Out', 'bowled', 'caught', 'lbw', 'stumped', 'run out',
       'caught and bowled', 'retired hurt', 'hit wicket', 'retired out',
       'obstructing the field'], dtype=object)

In [50]:
valid_methods = ['bowled', 'caught','lbw', 'stumped','caught and bowled', 'hit wicket']
df['bowler_wickets'] = df['Method'].isin(valid_methods).astype(int) #isin here it will checks the the valid_list values are there in the Metod if it is there it will return true else return false


In [56]:
df['is_dot_ball'] = ((df['Runs From Ball'] == 0) & (df['Valid Ball'] )).astype(int) #& for element wise or bit wise we will use '&' as and
